src/dispatch/dispatch_virtual.h

```diff
@@ -0,0 +1,185 @@
+// (NTESS). Under the terms of Contract DE-NA0003525 with NTESS, the U.S.
+// Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined(CHECKPOINT_DISPATCH_VIRTUAL_H)
+#define CHECKPOINT_DISPATCH_VIRTUAL_H
+
+#include "dispatch.h"
+
+#include <vector>
+#include <tuple>
+#include <functional>
+
+namespace serdes {
+  using AutoHandlerType = int;
+
+  struct SERIALIZE_CONSTRUCT_TAG {};
+
+  template <typename T>
+  using RegistryType = std::vector<
+    std::tuple<
+      int,
+      std::function<T*(void)>
+      >
+    >;
+
+  template <typename ObjT>
+  struct Registrar {
+    Registrar();
+    AutoHandlerType index;
+  };
+
+  template <typename ObjT>
+  struct Type {
+    static AutoHandlerType const idx;
+  };
+
+  template <typename T>
+  inline RegistryType<T>& getRegistry() {
+    static RegistryType<T> reg;
+    return reg;
+  }
+
+  template <typename ObjT>
+  Registrar<ObjT>::Registrar() {
+    using BaseType = typename ObjT::SerDerBaseType;
+
+    auto& reg = getRegistry<BaseType>();
+    index = reg.size();
+
+    debug_serdes("registrar: %ld, %s\n", reg.size(), typeid(ObjT).name());
+
+    reg.emplace_back(
+		     std::make_tuple(
+				     index,
+				     []() -> BaseType* { return new ObjT(SERIALIZE_CONSTRUCT_TAG{}); }
```

Comment:
This eventually should call into dispatch, with the `SERIALIZE_CONSTRUCT_TAG` overload added to that code. For now, what we have here is fine.
```C++
DeserializerDispatch<Unpacker, ObjT> dispatch;
dispatch(unpacker, buffer);
```